rust/kernel/hrtimer.rs

```diff
@@ -0,0 +1,279 @@
+    {
+        // SAFETY: By the safety requirement of this trait, the trait
+        // implementor will have a `Timer` field at the specified offset.
+        unsafe { (ptr as *mut u8).sub(Self::OFFSET) as *mut Self }
+    }
+}
+
+/// Implemented by pointer types that can be the target of a C timer callback.
+pub trait RawTimerCallback: RawTimer {
+    /// Callback to be called from C.
+    ///
+    /// # Safety
+    ///
+    /// Only to be called by C code in `hrtimer`subsystem.
+    unsafe extern "C" fn run(ptr: *mut bindings::hrtimer) -> bindings::hrtimer_restart;
+}
+
+/// Implemented by pointers to structs that can the target of a timer callback
+pub trait TimerCallback {
+    /// The type of the reciever of the callback
+    type Receiver<'a>: RawTimerCallback;
+
+    /// Called by the timer logic when the timer fires
+    fn run(this: Self::Receiver<'_>);
+}
+
+impl<T: Sync> RawTimer for Pin<&T>
+where
+    T: HasTimer<T>,
+{
+    fn schedule(self, expires: u64) {
+        // SAFETY: We are never moving the pointee of `self`
+        let unpinned = unsafe { Pin::into_inner_unchecked(self) };
+
+        let self_ptr = unpinned as *const T;
+
+        // SAFETY: `self_ptr` is a valid pointer to a `T`
+        let timer_ptr = unsafe { T::raw_get_timer(self_ptr) };
+
+        // `Timer` is `repr(transparent)`
+        let c_timer_ptr = timer_ptr.cast::<bindings::hrtimer>();
+
+        // Schedule the timer - if it is already scheduled it is removed and
+        // inserted
+
+        // SAFETY: c_timer$_ptr points to a valid hrtimer instance that was
+        // initialized by `hrtimer_init`
+        unsafe {
+            bindings::hrtimer_start_range_ns(
+                c_timer_ptr as *mut _,
+                expires as i64,
+                0,
+                bindings::hrtimer_mode_HRTIMER_MODE_REL,
+            );
+        }
+    }
+}
+
+impl<'a, T: Sync> RawTimerCallback for Pin<&'a T>
+where
+    T: HasTimer<T>,
+    T: TimerCallback<Receiver<'a> = Self> + 'a,
+{
+    unsafe extern "C" fn run(ptr: *mut bindings::hrtimer) -> bindings::hrtimer_restart {
+        // `Timer` is `repr(transparent)`
+        let timer_ptr = ptr.cast::<Timer<T>>();
+
+        // SAFETY: By C API contract ptr is the pointer we passed when enqueing
+        // the timer, so it is a `Timer<T>` embedded in a `T`
+        let receiver_ptr = unsafe { T::timer_container_of(timer_ptr) };
+
+        // SAFETY: The pointer was returned by `T::timer_container_of` so it points to a valid `T`
+        let receiver_ref = unsafe { &*receiver_ptr };
+
+        // SAFETY: We are not moving out of `receiver_pin`
+        let receiver_pin = unsafe { Pin::new_unchecked(receiver_ref) };
+
+        T::run(receiver_pin);
+
+        bindings::hrtimer_restart_HRTIMER_NORESTART
+    }
+}
+
+/// Derive the [`HasTimer`] trait for a struct that contains a field of type
+/// [`Timer`]. See the module level documentation for an example.
+#[macro_export]
+macro_rules! impl_has_timer {
```

Comment:
A small example here would be good, or link to the module docs for more information (since the macro will show up in the top-level docs)